stripe/src/main/java/com/stripe/android/net/PollingNetworkHandler.java

```diff
@@ -0,0 +1,188 @@
+package com.stripe.android.net;
+
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.Looper;
+import android.os.Message;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.stripe.android.exception.PollingFailedException;
+import com.stripe.android.exception.StripeException;
+import com.stripe.android.model.Source;
+import com.stripe.android.model.SourceRedirect;
+
+/**
+ * Class to handle polling on a background thread.
+ */
+class PollingNetworkHandler {
+
+    private static final String AUTHORIZATION_FAILED = "The redirect authorization request failed.";
+    private static final long DEFAULT_TIMEOUT_MS = 10000L;
+    private static final int INITIAL_DELAY_MS = 100;
+    private static final String POLLING_EXPIRED = "The polling request has expired.";
+    private static final int POLLING_MULTIPLIER = 2;
+
+    private static final int SUCCESS = 1;
+    private static final int PENDING = 2;
+    private static final int FAILURE = 3;
+    private static final int ERROR = -1;
+    private static final int EXPIRED = -2;
+
+    private final String mClientSecret;
+    private final String mPublishableKey;
+    private final String mSourceId;
+
+    private final long mTimeoutMs;
+    private final boolean mIsInSingleThreadMode;
+    @NonNull private final SourceRetriever mSourceRetriever;
+
+    private Handler mNetworkHandler;
+    private Handler mUiHandler;
+
+    private final Runnable pollRunnable = new Runnable() {
+        @Override
+        public void run() {
+            Message message = null;
+            try {
+                Source source = mSourceRetriever.retrieveSource(
+                        mSourceId,
+                        mClientSecret,
+                        mPublishableKey);
+                if (source.getRedirect() != null) {
+                    switch (source.getRedirect().getStatus()) {
+                        case SourceRedirect.SUCCEEDED:
+                            message = mUiHandler.obtainMessage(SUCCESS);
+                            break;
+                        case SourceRedirect.PENDING:
+                            message = mUiHandler.obtainMessage(PENDING);
+                            break;
+                        case SourceRedirect.FAILED:
+                            message = mUiHandler.obtainMessage(FAILURE);
+                            break;
+                    }
+                }
+            } catch (StripeException stripeEx) {
+                message = mUiHandler.obtainMessage(ERROR, stripeEx);
+            } finally {
+                if (message != null) {
+                    mUiHandler.sendMessage(message);
+                }
+            }
+        }
+    };
+
+    PollingNetworkHandler(@NonNull final String sourceId,
```

Comment:
I'm not totally equipped to read the behavior of this implementation, but it seems a bit different from what we do in the iOS SDK. I don't know what the platform-specific differences should be, but just to make sure we're doing a roughly equivalent thing here, this is a summary of what we do on iOS:

Each time we poll:
- First, we check if we've hit the timeout or the maximum number of retries
  - if so, complete with the last retrieved source and latest error
- Then, we fetch the source
  - If we receive a 4xx response:
    - stop polling and complete with the last retrieved source and an error
  - If we receive a 200 response:
    - Check the source's status
      - if it's `pending`, continue polling (without backing off)
      - if it's not `pending`, complete with the source
  - If we receive another status code:
    - something's wrong – retry with exponential backoff

https://github.com/stripe/stripe-ios/blob/master/Stripe/STPSourcePoller.m